subsys/net/lib/prometheus/formatter.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2024 Mustafa Abdullah Kus, Sparse Technology
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/net/prometheus/formatter.h>
+#include <zephyr/net/prometheus/collector.h>
+#include <zephyr/net/prometheus/metric.h>
+#include <zephyr/net/prometheus/histogram.h>
+#include <zephyr/net/prometheus/summary.h>
+#include <zephyr/net/prometheus/gauge.h>
+#include <zephyr/net/prometheus/counter.h>
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+#include <stddef.h>
+#include <errno.h>
+
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pm_formatter, CONFIG_PROMETHEUS_LOG_LEVEL);
+
+static int write_to_buffer(char *buffer, size_t buffer_size, const char *str)
+{
+	int ret;
+
+	/* helper function to write formatted string to buffer, handling buffer size */
+	ret = snprintf(buffer + strlen(buffer), buffer_size - strlen(buffer), "%s", str);
+	if (ret >= 0 && ret < buffer_size - strlen(buffer)) {
+		return ret;
+	}
+
+	/* error (insufficient buffer size) */
+	return 0;
+}
+
+static int write_metric_to_buffer(char *buffer, size_t buffer_size, const char *format, ...)
+{
+	/* helper function to write formatted metric to buffer */
+	char temp_buffer[CONFIG_PROMETHEUS_METRIC_FORMAT_BUFFER_SIZE];
+	va_list args;
+
+	va_start(args, format);
+	vsnprintf(temp_buffer, sizeof(temp_buffer), format, args);
+	va_end(args);
+
+	return write_to_buffer(buffer, buffer_size, temp_buffer);
```

Comment:
I think it is possible to combine the `vsnprintf()` from here and the `snprintf()` from `write_to_buffer()` which is used only once in the end...

```suggestion
	va_list args;
	size_t len = strlen(buffer);

	if (len >= buffer_size) {
		return -ENOMEM;
	}
	buffer += len;
	buffer_size -= len;

	va_start(args, format);
	len = vsnprintf(buffer, buffer_size, format, args);
	va_end(args);
	if (len >= buffer_size) {
		return -ENOMEM;
	}

	return 0;
```

This would allow to remove `CONFIG_PROMETHEUS_METRIC_FORMAT_BUFFER_SIZE` from Kconfig.
This is `>=` instead of `>` because `vsnprintf()` does not count the terminating `'\0'`.
This also always return `-error_number` on error and `0` on success.
That is just a small optimization though...

Reply:
It is good suggestion for that case. I did applied :)